File: templates/nodejs-service/service/src/transport/rmqEventMap.ts
import { RmqEventConfig } from "./interfaces";

export class RmqEventMap {
    constructor(private map: {[eventName: string]: RmqEventConfig}) {}

    public getExchangeName(eventName: string): string {
        if (eventName in this.map && "exchangeName" in this.map[eventName] && this.map[eventName].exchangeName !== "") {
            return this.map[eventName].exchangeName;
        }
        return eventName;
    }

    public getQueueName(eventName: string): string {
        if (eventName in this.map && "queueName" in this.map[eventName] && this.map[eventName].queueName !== "") {
            return this.map[eventName].queueName;
        }
        return eventName;
    }

    public getDeadLetterExchange(eventName: string): string {
        if (eventName in this.map && "deadLetterExchange" in this.map[eventName] && this.map[eventName].deadLetterExchange !== "") {
            return this.map[eventName].deadLetterExchange;
        }
        const exchangeName = this.getExchangeName(eventName);
        return `${exchangeName}.dlx`;
    }

    public getDeadLetterQueue(eventName: string): string {
        if (eventName in this.map && "deadLetterQueue" in this.map[eventName] && this.map[eventName].deadLetterQueue !== "") {
            return this.map[eventName].deadLetterQueue;
        }
        const queueName = this.getQueueName(eventName);
        return `${queueName}.dlx`;
    }

    public getTtl(eventName: string): number {
        if (eventName in this.map && "ttl" in this.map[eventName] && this.map[eventName].ttl > 0) {
            return this.map[eventName].ttl;
        }
        return 0;
    }

    public getRpcTimeout(eventName: string): number {
        if (eventName in this.map && "rpcTimeout" in this.map[eventName] && this.map[eventName].rpcTimeout as number > 0) {
            return this.map[eventName].rpcTimeout as number;
        }
        return 30000;
    }

    public getAutoAck(eventName: string): boolean {
        if (eventName in this.map && "autoAck" in this.map[eventName]) {
            return this.map[eventName].autoAck;
        }
        return false;
    }


}

File: templates/go-service/service/transport/rmqEventMap.go
package transport

import (
	"fmt"
	"time"

	"github.com/streadway/amqp"
)

// RmqEventConfig single event config
type RmqEventConfig struct {
	QueueName          string
	ExchangeName       string
	RPCTimeout         int
	DeadLetterExchange string
	DeadLetterQueue    string
	TTL                int
	AutoAck            bool
}

// RmqEventMap map of event config
type RmqEventMap map[string]RmqEventConfig

// GetExchangeName based on event
func (m *RmqEventMap) GetExchangeName(eventName string) (exchangeName string) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok && config.ExchangeName != "" {
		return config.ExchangeName
	}
	return eventName
}

// GetQueueName based on event
func (m *RmqEventMap) GetQueueName(eventName string) (queueName string) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok && config.QueueName != "" {
		return config.QueueName
	}
	return eventName
}

// GetDeadLetterExchange based on event
func (m *RmqEventMap) GetDeadLetterExchange(eventName string) (deadLetterExchangeName string) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok && config.DeadLetterExchange != "" {
		return config.DeadLetterExchange
	}
	return fmt.Sprintf("%s.dlx", m.GetExchangeName(eventName))
}

// GetDeadLetterQueue based on event
func (m *RmqEventMap) GetDeadLetterQueue(eventName string) (deadLetterQueueName string) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok && config.DeadLetterQueue != "" {
		return config.DeadLetterExchange
	}
	return fmt.Sprintf("%s.dlx", m.GetQueueName(eventName))
}

// GetTTL based on event
func (m *RmqEventMap) GetTTL(eventName string) (TTL int32) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok && config.TTL > 0 {
		return int32(config.TTL)
	}
	return int32(0)
}

// GetQueueArgs based on event
func (m *RmqEventMap) GetQueueArgs(eventName string) (args amqp.Table) {
	args = make(amqp.Table)
	if m.GetTTL(eventName) <= 0 {
		return nil
	}
	args["x-dead-letter-exchange"] = m.GetDeadLetterExchange(eventName)
	args["x-message-ttl"] = m.GetTTL(eventName)
	return args
}

// GetRPCTimeout based on event
func (m *RmqEventMap) GetRPCTimeout(eventName string) (RPCTimeout time.Duration) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok && config.RPCTimeout > 0 {
		return time.Duration(config.RPCTimeout)
	}
	return 30000
}

// GetAutoAck based on event
func (m *RmqEventMap) GetAutoAck(eventName string) (autoAck bool) {
	eventMap := *m
	if config, ok := eventMap[eventName]; ok {
		return config.AutoAck
	}
	return false
}
